### ISSUE TITLE
✨ Default platform to buildkitd's platform

### ISSUE THREAD

**Issue by aluzzardi** [dagger/dagger#1996]
I think it's performance critical to do this ASAP. I know we have unfinished discussions regarding the DX for multi-platform and how to override one.

I propose as a stopgap to add a `--platform` flag to `dagger do` for overrides -- no CUE syntax.

Background:

When the requested architecture (e.g. `linux/amd64` -- which is our default) is different from buildkit's architecture (e.g. `linux/aarch64` -- which is the actual architecture of my MacBook), buildkit uses `qemu` for emulation. TLDR: it's 10x slower.

`shellcheck` with `linux/amd64` takes 14s to run on my laptop:

```
$ time dagger do lint shell                                                                                                                                                                                                                                                                                                                         [✔] actions.lint.shell                                                                                                                                                                                                                                                                                                                                                            1.0s
[✔] client.filesystem."../".read                                                                                                                                                                                                                                                                                                                                                  0.9s
[✗] actions.lint.shell.container                                                                                                                                                                                                                                                                                                                                                 12.6s
#5 11.26                                                                                                                                                                                                                                                                                                                                                                              
#5 11.26 In ./tests/shellcheck.sh line 1:                                                                                                                                                                                                                                                                                                                                             
                                                  
dagger do lint shell  1.06s user 0.39s system 10% cpu 14.226 total
```

`shellcheck` with `linux/aarch64` (native platform) runs in 1.4s:
```
$ time dagger do lint shell
[✔] actions.lint.shell                                                                                                                                                                                                                                                                                                                                                            0.4s
[✔] client.filesystem."../".read                                                                                                                                                                                                                                                                                                                                                  0.3s
[✗] actions.lint.shell.container                                                                                                                                                                                                                                                                                                                                                  0.4s
#5 0.346                                                                                                                                                                                                                                                                                                                                                                              
#5 0.346 In ./tests/shellcheck.sh line 1:                                                                                                                                                                                                                                                                                                                                             

dagger do lint shell  1.01s user 0.28s system 90% cpu 1.417 total
```

/cc @helderco @shykes @samalba @TomChv

**Comment by marcosnils** [comment]
:wave: if we agree on the `--platform` flag I can help shipping this one.

**Comment by shykes** [comment]
* Changing the default from "always linux/amd64 no matter what" to "whatever buildkitd wants to default to" makes sense to me. Given the performance hit it seems like a no-brainer to do this right away.

* For the `--platform` flag: can you remind me what are the options today for explicitly setting a platform? Is there already one, and we would be changing how it works? Or would we be introducing this feature for the first time?

**Comment by aluzzardi** [comment]
> For the --platform flag: can you remind me what are the options today for explicitly setting a platform? Is there already one, and we would be changing how it works? Or would we be introducing this feature for the first time?

There's a `platform` field in `#Plan` to set the default platform. This was part of the broader multi-platform effort which is stuck at the design stage. I suggest removing the `platform` field from `#Plan` and moving it to a flag (a la `docker build`) so we don't have to wait to figure out what the CUE DX for multi-platform is.

**Comment by marcosnils** [comment]
> I suggest removing the `platform` field from `#Plan` and moving it to a flag (a la `docker build`) so we don't have to wait to figure out what the CUE DX for multi-platform is.

Well.. the `--platform` flag could be set today by using the `--with` flag, correct? I guess we agree it's no the best UX but maybe we can fallback to that until we find the time to think about the "idea" initial multi-platform solution. In any case, the automatic buildkit target platform detection needs to be implemented regardless of the flag. Everyone ok for me taking this one?

**Comment by aluzzardi** [comment]
> Everyone ok for me taking this one?

Yes, go for it! You'll find pointers in the buildx code base, maybe there's some helper code we can re-import (we already depend on buildx).

TBH I don't exactly remember why we have to detect the platform and pass it to buildkit, rather than not specifying it at all. I think the buildkit API requires a platform no matter what, but I'm not 100% sure.

**Comment by TomChv** [comment]
> TBH I don't exactly remember why we have to detect the platform and pass it to buildkit, rather than not specifying it at all. I think the buildkit API requires a platform no matter what, but I'm not 100% sure.

No, by default buildkit will take the **host platform**, but that's not a good solution because `darwin/arm64` (m1 mac) isn't available in docker image.
So we did 2 things : 
- Force platform by default to be `linux/amd64`
- Add a platform field to customise platform if user wants it

> Use `--platform`

It's a good idea but only temporary, that will be deprecated when we will integrate multi-platform or find a good Cue DX. 
Try to use keep integration simple so it will be easy to revert.

As well, the best things to do would be to retrieve the daemon worker's architecture and pass it by default.
I did some research when I worked on multi-platform but I left that part unfinished because I had other stuff to do.

So don't hesitate to ping me @marcosnils if you need anything, I dug a lot into buildkit and platform.

**Comment by shykes** [comment]
My preference would be to remove `--platform` for now (and also remove the top-level `platform` field which I thought we had already removed). And only ship the change of default (if everyone agrees on that part).

It's ambiguous what it means to "specify the platform". Does it change the platform on containers run by Dagger? The platform of images built and pushed by Dagger? Both? Concepts from regular container build tools don't translate directly to Dagger, because it does more than build.

IMO `--platform` and top-level `platform` are already a beginning of multi-platform design, and as I said before, I'm worried about shipping a piece of a design which we haven't finished. We may have to go back and break it later - better to not ship it at all.

**Comment by TomChv** [comment]
> And only ship the change of default (if everyone agrees on that part).

I agree but then users will not be able to run nor push container/image in another architecture than their host one (or daemon worker's depending on what marcos will do)

 > It's ambiguous what it means to "specify the platform". Does it change the platform on containers run by Dagger? The platform of images built and pushed by Dagger? Both? 

It's both

> IMO --platform and top-level platform are already a beginning of multi-platform design

I disagree, being able to specify in which platform `dagger do` is important,  some images are not available in `linux/arm64` ([`mysql:latest`](https://hub.docker.com/_/mysql?tab=tags) for example) so it can be a limitation for mac users in certain cases. Same if you want to push a docker image in a different platform than your host's

Indeed, we can also remove `--platform` for now and wait for a user to report that issue when he will meet it, but we can also choose to solve that problem with a temporary flag until we find a good solution.

**Comment by shykes** [comment]
> > And only ship the change of default (if everyone agrees on that part).
> 
> I agree but then users will not be able to run nor push container/image in another architecture than their host one (or daemon worker's depending on what marcos will do)
> 
> > It's ambiguous what it means to "specify the platform". Does it change the platform on containers run by Dagger? The platform of images built and pushed by Dagger? Both?
> 
> It's both
> 
> > IMO --platform and top-level platform are already a beginning of multi-platform design
> 
> I disagree, being able to specify in which platform `dagger do` is important, some images are not available in `linux/arm64` ([`mysql:latest`](https://hub.docker.com/_/mysql?tab=tags) for example) so it can be a limitation for mac users in certain cases. Same if you want to push a docker image in a different platform than your host's

I didn't say it wasn't important. I said that defining a `--platform` flag or a top-level `platform` setting is a major design decision for multi-platform support in Dagger. Which is why I wanted that top-level `platform` field to be removed before release, and why now I would like that `--platform` flag removed. I don't want us to make such a major decision by default, without even realizing it.

> Indeed, we can also remove `--platform` for now and wait for a user to report that issue when he will meet it, but we can also choose to solve that problem with a temporary flag until we find a good solution.

"No is temporary, yes is forever". The only temporary option here is to remove the flag/setting. If we keep either, it will be very painful to get rid of it.

**Comment by aluzzardi** [comment]
> My preference would be to remove --platform for now (and also remove the top-level platform field which I thought we had already removed). And only ship the change of default (if everyone agrees on that part).

If we do that, then I won't be able to deploy from my laptop

**Comment by TomChv** [comment]
> If we do that, then I won't be able to deploy from my laptop

Why? If we set platform to `linux/arm64` for mac m1, that should works no? Are you using something special?

**Comment by aluzzardi** [comment]
> Why? If we set platform to linux/arm64 for mac m1, that should works no? Are you using deploying special?

It will produce `linux/arm64` artifacts which won't be able to run on most infrastructure providers (GKE/ECS/Cloud Run/etc)

**Comment by samalba** [comment]
> > Why? If we set platform to linux/arm64 for mac m1, that should works no? Are you using deploying special?
> 
> It will produce `linux/arm64` artifacts which won't be able to run on most infrastructure providers (GKE/ECS/Cloud Run/etc)

I would default to the os: `linux` and arch: `client's arch` for performance reason, with the option to override the arch for a specific action for cross-builds. A global CLI flag to override the platform globally then becomes a nice to have.

**Comment by marcosnils** [comment]
Closed by #2028
